### PR TITLE
[Merged by Bors] - Support tls authentication

### PR DESCRIPTION
**Pull request by maltesander**
# Description

Implemented support for:
- Client TLS (defaults to "tls")
- Authenticated client TLS (optional)
- Internal broker TLS (defaults to "tls)

Sorry for the big PR :(

closes https://github.com/stackabletech/kafka-operator/issues/12

## Review Checklist

- [x] Code contains useful comments
- [x] CRD change approved (or not applicable)
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: rust/operator/src/kafka_controller.rs [hunk 1]
```diff
@@ -741,3 +865,76 @@ fn build_broker_rolegroup_statefulset(
...
+            protocol: Some("TCP".to_string()),
+            ..ContainerPort::default()
+        });
+    }
+
+    ports
+}
+
+fn create_tls_volume(volume_name: &str, tls_secret_class: Option<&TlsSecretClass>) -> Volume {
+    let mut secret_class_name = "tls";
```
Comment (sbernauer) [inline rust/operator/src/kafka_controller.rs:926]: We could use something like `let secret_class_name = tls_secret_class.and_then(|t| t.secret_class).unwrap_or("tls")` (not tested)

**Review by sbernauer — CHANGES_REQUESTED** [review]
Reviewed the code so far, now I'm going to look into the tests

File: tests/templates/kuttl/tls/10-install-zookeeper.yaml.j2 [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    tls:
+      secretClass: tls
+    clientAuthentication:
+      authenticationClass: zk-client-tls
+{% else %}
+    tls: null
+{% endif %}
+{% if test_scenario['values']['use-internal-tls'] == 'true' %}
+    internalTlsSecretClass: tls
+{% endif %}
```
Comment (sbernauer) [inline tests/templates/kuttl/tls/10-install-zookeeper.yaml.j2:19]: Currently connection Kafka to tls enabled zk is not supported 
```suggestion
  config:
    tls: null
```

File: rust/operator/src/discovery.rs [hunk 3]
```diff
@@ -46,13 +48,20 @@ pub async fn build_discovery_configmaps(
     svc: &Service,
 ) -> Result<Vec<ConfigMap>, Error> {
     let name = owner.name();
+    let port_name = if kafka.client_tls_secret_class().is_some()
+        || kafka.client_authentication_class().is_some()
+    {
+        SECURE_CLIENT_PORT_NAME
+    } else {
+        CLIENT_PORT_NAME
+    };
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator/src/discovery.rs:57]: done

File: rust/operator/src/kafka_controller.rs [hunk 4]
```diff
@@ -181,6 +253,38 @@ pub async fn reconcile_kafka(kafka: Arc<KafkaCluster>, ctx: Arc<Ctx>) -> Result<
...
 
+    let client_authentication_class = if let Some(auth_class) = kafka.client_authentication_class()
+    {
+        Some(
+            client
+                .get::<AuthenticationClass>(auth_class, None) // AuthenticationClass has ClusterScope
+                .await
+                .context(AuthenticationClassRetrievalSnafu {
+                    authentication_class: ObjectRef::<AuthenticationClass>::new(auth_class),
+                })?,
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator/src/kafka_controller.rs:265]: Not in the used operator-rs version, i would leave that for another pr, added a TODO.

File: rust/operator/src/kafka_controller.rs (hunk 1 above)
Comment (maltesander) [reply rust/operator/src/kafka_controller.rs]: done

**Comment by sbernauer** [comment]
Starting review

File: tests/templates/kuttl/tls/20-install-kafka.yaml.j2 [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+{% if test_scenario['values']['use-client-tls'] == 'true' %}
+    tls:
+      secretClass: tls
+{% else %}
+    tls: null
+{% endif %}
+{% if test_scenario['values']['use-client-auth-tls'] == 'true' %}
+    clientAuthentication:
+      authenticationClass: test-kafka-client-auth-tls
+{% endif %}
```
Comment (sbernauer) [inline tests/templates/kuttl/tls/20-install-kafka.yaml.j2:66]: Just for consistency
```suggestion
{% if test_scenario['values']['use-client-auth-tls'] == 'true' %}
    clientAuthentication:
      authenticationClass: test-kafka-client-auth-tls
{% else %}
    clientAuthentication: null
{% endif %}
```

File: tests/templates/kuttl/tls/30-prepare-test-kafka.yaml.j2 [hunk 6]
```diff
@@ -0,0 +1,11 @@
+---
+apiVersion: kuttl.dev/v1beta1
+kind: TestStep
+commands:
+{% if test_scenario['values']['use-client-auth-tls'] == 'true' %}
+  - script: kubectl cp -n $NAMESPACE ./test_client_auth_tls.sh test-kafka-broker-default-0:/tmp
+  - script: kubectl cp -n $NAMESPACE ./wrong_keystore.p12 test-kafka-broker-default-0:/tmp
```
Comment (sbernauer) [inline tests/templates/kuttl/tls/30-prepare-test-kafka.yaml.j2:7]: What's the lifetime of the certs in the keystore?

**Review by sbernauer — COMMENTED** [review]
Reviewed the tests. The number of tests exploded, but IMHO out of scope for this PR

File: tests/templates/kuttl/tls/30-prepare-test-kafka.yaml.j2 (hunk 6 above)
Comment (maltesander) [reply tests/templates/kuttl/tls/30-prepare-test-kafka.yaml.j2:7]: yeah good point, generated new ones:
```
openssl x509 -enddate -noout -in ca.crt 
notAfter=Jul 12 09:21:36 2122 GMT
```

**Review by sbernauer — COMMENTED** [review]
The custom tests took 13h and failed :see_no_evil: 
https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/kafka-operator-it-custom/9/

**Review by sbernauer — APPROVED** [review]
Please wait for https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/kafka-operator-it-custom/10/console

**Comment by sbernauer** [comment]
Next try https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/kafka-operator-it-custom/11/console

**Comment by maltesander** [comment]
bors merge

**Pull request closed without merging**